"Buffer full" when saving a snapshot

This error was received when saving a large snapshot:
```
Caused by: java.lang.IllegalStateException: buffer full
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl$CustomByteArrayOutputStream.ensureCapacity(AsyncSnapshotWriterImpl.java:464)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl$CustomByteArrayOutputStream.write(AsyncSnapshotWriterImpl.java:452)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.getBufferContentsAndClear(AsyncSnapshotWriterImpl.java:206)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.lambda$flushPartition$1(AsyncSnapshotWriterImpl.java:198)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.putAsyncToMap(AsyncSnapshotWriterImpl.java:231)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.flushPartition(AsyncSnapshotWriterImpl.java:198)
    at com.hazelcast.jet.impl.util.AsyncSnapshotWriterImpl.offer(AsyncSnapshotWriterImpl.java:170)
    at com.hazelcast.jet.impl.execution.StoreSnapshotTasklet.addToInbox(StoreSnapshotTasklet.java:145)
    at com.hazelcast.jet.impl.execution.ConcurrentInboundEdgeStream$ItemDetector.test(ConcurrentInboundEdgeStream.java:233)
    at com.hazelcast.internal.util.concurrent.OneToOneConcurrentArrayQueue.drain(OneToOneConcurrentArrayQueue.java:92)
    at com.hazelcast.jet.impl.execution.ConcurrentInboundEdgeStream.drainQueue(ConcurrentInboundEdgeStream.java:190)
    at com.hazelcast.jet.impl.execution.ConcurrentInboundEdgeStream.drainTo(ConcurrentInboundEdgeStream.java:108)
    at com.hazelcast.jet.impl.execution.StoreSnapshotTasklet.stateMachineStep(StoreSnapshotTasklet.java:89)
    at com.hazelcast.jet.impl.execution.StoreSnapshotTasklet.call(StoreSnapshotTasklet.java:74)
    at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:302)
    at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.lambda$run$0(TaskletExecutionService.java:278)
    at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:803)
    at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:283)
    at java.base/java.lang.Thread.run(Thread.java:834)
```

Credits: @mtishchuk

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImpl.java
@@ -51,7 +51,7 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
 
     public static final int DEFAULT_CHUNK_SIZE = 128 * 1024;
 
-    final int usableChunkSize; // this includes the serialization header for byte[], but not the terminator
+    final int usableChunkCapacity; // this includes the serialization header for byte[], but not the terminator
     final byte[] serializedByteArrayHeader = new byte[3 * Bits.INT_SIZE_IN_BYTES];
     final byte[] valueTerminator;
     final AtomicInteger numConcurrentAsyncOps;
@@ -67,7 +67,7 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
     private final SnapshotContext snapshotContext;
     private final String vertexName;
     private final int memberCount;
-    private IMap currentMap;
+    private IMap<SnapshotDataKey, Object> currentMap;
     private long currentSnapshotId;
     private final AtomicReference<Throwable> firstError = new AtomicReference<>();
     private final AtomicInteger numActiveFlushes = new AtomicInteger();
@@ -87,6 +87,9 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
     // for test
     AsyncSnapshotWriterImpl(int chunkSize, NodeEngine nodeEngine, SnapshotContext snapshotContext,
                             String vertexName, int memberIndex, int memberCount) {
+        if (Integer.bitCount(chunkSize) != 1) {
+            throw new IllegalArgumentException("chunkSize must be a power of two, but is " + chunkSize);
+        }
         this.nodeEngine = nodeEngine;
         this.partitionService = nodeEngine.getPartitionService();
         this.logger = nodeEngine.getLogger(getClass());
@@ -101,12 +104,7 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
         Bits.writeInt(serializedByteArrayHeader, Bits.INT_SIZE_IN_BYTES, SerializationConstants.CONSTANT_TYPE_BYTE_ARRAY,
                 useBigEndian);
 
-        buffers = new CustomByteArrayOutputStream[partitionService.getPartitionCount()];
-        for (int i = 0; i < buffers.length; i++) {
-            buffers[i] = new CustomByteArrayOutputStream(chunkSize);
-            buffers[i].write(serializedByteArrayHeader, 0, serializedByteArrayHeader.length);
-        }
-
+        buffers = createAndInitBuffers(chunkSize, partitionService.getPartitionCount(), serializedByteArrayHeader);
         JetService jetService = nodeEngine.getService(JetService.SERVICE_NAME);
         this.partitionKeys = jetService.getSharedPartitionKeys();
         this.partitionSequence = memberIndex;
@@ -117,20 +115,37 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
                 SnapshotDataValueTerminator.INSTANCE).toByteArray();
         valueTerminator = Arrays.copyOfRange(valueTerminatorWithHeader, HeapData.TYPE_OFFSET,
                 valueTerminatorWithHeader.length);
-        usableChunkSize = chunkSize - valueTerminator.length;
+        usableChunkCapacity = chunkSize - valueTerminator.length - serializedByteArrayHeader.length;
+        if (usableChunkCapacity <= 0) {
+            throw new IllegalArgumentException("too small chunk size: " + chunkSize);
+        }
+    }
+
+    private static CustomByteArrayOutputStream[] createAndInitBuffers(
+            int chunkSize,
+            int partitionCount,
+            byte[] serializedByteArrayHeader
+    ) {
+        CustomByteArrayOutputStream[] buffers = new CustomByteArrayOutputStream[partitionCount];
+        for (int i = 0; i < buffers.length; i++) {
+            buffers[i] = new CustomByteArrayOutputStream(chunkSize);
+            buffers[i].write(serializedByteArrayHeader, 0, serializedByteArrayHeader.length);
+        }
+        return buffers;
     }
 
     private void consumePutResponse(Object response, Throwable throwable) {
+        try {
+            assert response == null : "put operation overwrote a previous value: " + response;
+        } catch (AssertionError e) {
+            throwable = e;
+        }
         if (throwable != null) {
             logger.severe("Error writing to snapshot map", throwable);
             firstError.compareAndSet(null, throwable);
-            numActiveFlushes.decrementAndGet();
-            numConcurrentAsyncOps.decrementAndGet();
-        } else {
-            assert response == null : "put operation overwrote a previous value: " + response;
-            numActiveFlushes.decrementAndGet();
-            numConcurrentAsyncOps.decrementAndGet();
         }
+        numActiveFlushes.decrementAndGet();
+        numConcurrentAsyncOps.decrementAndGet();
     }
 
     @Override
@@ -139,9 +154,9 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
         int partitionId = partitionService.getPartitionId(entry.getKey());
         int length = entry.getKey().totalSize() + entry.getValue().totalSize() - 2 * HeapData.TYPE_OFFSET;
 
-        // if single entry is larger than usableChunkSize, send it alone. We avoid adding it to the ByteArrayOutputStream,
-        // since it will grow beyond maximum capacity and never shrink again.
-        if (length > usableChunkSize) {
+        // if the entry is larger than usableChunkSize, send it in its own chunk. We avoid adding it to the
+        // ByteArrayOutputStream since it would expand it beyond its maximum capacity.
+        if (length > usableChunkCapacity) {
             return putAsyncToMap(partitionId, () -> {
                 byte[] data = new byte[serializedByteArrayHeader.length + length + valueTerminator.length];
                 totalKeys++;
@@ -164,14 +179,15 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
             });
         }
 
-        // if the buffer will exceed usableChunkSize after adding this entry, flush it first
-        if (buffers[partitionId].size() + length > usableChunkSize && !flushPartition(partitionId)) {
+        // if the buffer after adding this entry and terminator would exceed the capacity limit, flush it first
+        CustomByteArrayOutputStream buffer = buffers[partitionId];
+        if (buffer.size() + length + valueTerminator.length > buffer.capacityLimit && !flushPartition(partitionId)) {
             return false;
         }
 
         // append to buffer
-        writeWithoutHeader(entry.getKey(), buffers[partitionId]);
-        writeWithoutHeader(entry.getValue(), buffers[partitionId]);
+        writeWithoutHeader(entry.getKey(), buffer);
+        writeWithoutHeader(entry.getValue(), buffer);
         totalKeys++;
         return true;
     }
@@ -233,7 +249,7 @@ public class AsyncSnapshotWriterImpl implements AsyncSnapshotWriter {
                     new SnapshotDataKey(partitionKeys[partitionId], currentSnapshotId, vertexName, partitionSequence),
                     data).toCompletableFuture();
             partitionSequence += memberCount;
-            future.whenCompleteAsync(putResponseConsumer);
+            future.whenComplete(putResponseConsumer);
             numActiveFlushes.incrementAndGet();
         } catch (HazelcastInstanceNotActiveException ignored) {
             return false;

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImplTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncSnapshotWriterImplTest.java
@@ -55,6 +55,7 @@ import java.util.stream.Stream;
 
 import static com.hazelcast.jet.Util.entry;
 import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+import static java.util.Collections.nCopies;
 import static java.util.stream.Collectors.joining;
 import static java.util.stream.Stream.generate;
 import static junit.framework.TestCase.assertEquals;
@@ -73,6 +74,7 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
     @Rule
     public ExpectedException exception = ExpectedException.none();
 
+    private NodeEngineImpl nodeEngine;
     private AsyncSnapshotWriterImpl writer;
     private IMap<SnapshotDataKey, byte[]> map;
     private InternalSerializationService serializationService;
@@ -89,7 +91,7 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
               .setImplementation(new AlwaysFailingMapStore());
 
         JetInstance instance = createJetMember(jetConfig);
-        NodeEngineImpl nodeEngine = ((HazelcastInstanceImpl) instance.getHazelcastInstance()).node.nodeEngine;
+        nodeEngine = ((HazelcastInstanceImpl) instance.getHazelcastInstance()).node.nodeEngine;
         serializationService = ((HazelcastInstanceImpl) instance.getHazelcastInstance()).getSerializationService();
         partitionService = nodeEngine.getPartitionService();
         snapshotContext = mock(SnapshotContext.class);
@@ -98,7 +100,7 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
         writer = new AsyncSnapshotWriterImpl(128, nodeEngine, snapshotContext, "vertex", 0, 1);
         when(snapshotContext.currentSnapshotId()).thenReturn(1L); // simulates starting new snapshot
         map = instance.getHazelcastInstance().getMap("map1");
-        assertTrue(writer.usableChunkSize > 0);
+        assertTrue(writer.usableChunkCapacity > 0);
     }
 
     @After
@@ -106,6 +108,20 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
         assertTrue(writer.flushAndResetMap());
         assertTrueEventually(() -> assertFalse(uncheckCall(() -> writer.hasPendingAsyncOps())));
         assertTrue(writer.isEmpty());
+    }
+
+    @Test
+    public void test_flushingAtEdgeCases() {
+        for (int i = 64; i < 196; i++) {
+            when(snapshotContext.currentMapName()).thenReturn(randomMapName());
+            writer = new AsyncSnapshotWriterImpl(128, nodeEngine, snapshotContext, "vertex", 0, 1);
+            try {
+                assertTrue(writer.offer(entry(serialize("k"), serialize(String.join("", nCopies(i, "a"))))));
+                assertTrue(writer.flushAndResetMap());
+            } catch (Exception e) {
+                throw new RuntimeException("error at i=" + i, e);
+            }
+        }
     }
 
     @Test
@@ -126,7 +142,8 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
     public void when_chunkSizeWouldExceedLimit_then_flushedAutomatically() {
         // When
         Entry<Data, Data> entry = entry(serialize("k"), serialize("v"));
-        int entriesInChunk = (writer.usableChunkSize - writer.serializedByteArrayHeader.length) / serializedLength(entry);
+        int entriesInChunk =
+                (writer.usableChunkCapacity - writer.serializedByteArrayHeader.length) / serializedLength(entry);
         assertTrue("entriesInChunk=" + entriesInChunk, entriesInChunk > 1 && entriesInChunk < 10);
 
         for (int i = 0; i < entriesInChunk; i++) {
@@ -177,7 +194,7 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
         String key = "k";
         String value = generate(() -> "a").limit(128).collect(joining());
         Entry<Data, Data> entry = entry(serialize(key), serialize(value));
-        assertTrue("entry not longer than usable chunk size", serializedLength(entry) > writer.usableChunkSize);
+        assertTrue("entry not longer than usable chunk size", serializedLength(entry) > writer.usableChunkCapacity);
         assertTrue(writer.offer(entry));
 
         // Then
@@ -201,7 +218,8 @@ public class AsyncSnapshotWriterImplTest extends JetTestSupport {
         // artificially increase number of async ops so that the writer cannot proceed
         writer.numConcurrentAsyncOps.set(JetService.MAX_PARALLEL_ASYNC_OPS);
         Entry<Data, Data> entry = entry(serialize("k"), serialize("v"));
-        int entriesInChunk = (writer.usableChunkSize - writer.serializedByteArrayHeader.length) / serializedLength(entry);
+        int entriesInChunk =
+                (writer.usableChunkCapacity - writer.serializedByteArrayHeader.length) / serializedLength(entry);
         assertTrue("entriesInChunk=" + entriesInChunk, entriesInChunk > 1 && entriesInChunk < 10);
         for (int i = 0; i < entriesInChunk; i++) {
             assertTrue(writer.offer(entry));

